<html>
<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<title>Frequently Asked Questions (PAML)</title>
<style><!--
.Normal
	{font-size:11.0pt;
	font-family:Palatino;}
.MsoToc1
	{font-size:11.0pt;
	font-family:Palatino;}
.MsoToc2
	{font-size:11.0pt;
	font-family:Palatino;}
.MsoHyperlink
	{color:blue;
	text-decoration:underline;
	text-underline:single;}
-->
</style>
</head>
<body lang=EN-GB link=blue vlink=purple class="Normal" bgcolor="#FFFFFF">
<p style='line-height:200%'><b><span lang=EN-US
style='font-size:18.0pt;color:red'>Frequently Asked Questions (PAML)</span></b></p>
<p style='line-height:200%'><span lang=EN-US>Ziheng Yang (July 2001)</span></p>
<p class=MsoToc1> <span class=MsoHyperlink><a href="#_Toc520201490">Compiling 
  the program and preparing files</a></span></p>
<blockquote> 
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201491">How do I compile the programs?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201492">Does paml read phylip format?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201493">Does paml read nexus (PAUP) format?</a></span></p>
</blockquote>
<p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201494">Windows and UNIX basics</a></span></p>
<p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201495">CODEML</a></span></p>
<blockquote>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201496">How can I get codeml to accept more that one data set at 
    a time?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201497">What are the limits in terms of number of sequences in the 
    alignment?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201498">Are my sequences too divergent for a sensible ML analysis?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201499">How do I test positive selection along specific lineages?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201500">Can I conclude positive selection if I found significantly 
    higher dN/dS ratios for my branches of interest than for other branches?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201501">What is the effect of assuming the same selective pressure 
    among lineages?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201502">Can I obtain a list of probabilities for all sites and all 
    categories under a NSsites model?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201503">The different NSsites models identified different sites.  
    Which model should I believe?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201504">Can I specify certain branches in a tree file to have independent 
    w parameters?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201505">Can I specify certain sites in an alignment to have independent 
    w parameters?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201506">Can I estimate dN and dS between all pairs of taxa in an 
    alignment?</a></span></p>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201507">Why do the inferred site classes not match the estimated 
    proportions under the NSsites models?</a></span></p>
</blockquote>
<p class=MsoToc1><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201508">BASEML</a></span></p>
<p class=MsoToc1><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201509">EVOLVER</a></span></p>
<blockquote>
  <p><span lang=EN-US style='font-size:10.0pt'><a
href="#_Toc520201510">Where is the control file for evolver?</a></span></p>
</blockquote>
<h1><a name="_Toc520201490"><span lang=EN-US>Compiling the program and preparing 
  files</span></a></h1>
<h2><a name="_Toc520201491"><span lang=EN-US>How do I compile the programs?</span></a></h2>
<p>Executables for Windows 95/98/Nt/2K are provided, as well as those for PowerMac 
  or G3/4.  The source codes are in ANSI C.  If you need to compile the programs 
  yourself.  Look at the makefile as well as the readme.txt file in the paml/src/ 
  directory.</p>
<h2><a name="_Toc520201492"><span lang=EN-US>Does paml read phylip format?</span></a></h2>
<p>Yes, paml programs read both the sequential and interlaved formats used in 
  phylip.  However there are the following complications.  First, paml requires 
  that you have the option variable I on the first line if the interlaved format 
  is used.  Second, phylip assumes 10 characters in a sequence name, while paml 
  assumes 30 characters (I think).  My suggestion is that you make sure to separate 
  the sequence name from the sequence by at least two spaces, in which case both 
  paml and phylip programs can read the same data file with no problem.  For example 
  the following works in both paml and phylip.</p>
<p><span lang=EN-US>4 20</span></p>
<p>sequence_1  TCATT CTATC TATCG TGATG</p>
<p>sequence_2  TCATT CTATC TATCG TGATG</p>
<p>sequence_3  TCATT CTATC TATCG TGATG</p>
<p>sequence_4  TCATT CTATC TATCG TGATG</p>
<p><span lang=EN-US>The following works with phylip but not paml.</span></p>
<p><span lang=EN-US>4 20</span></p>
<p><span lang=EN-US>sequence_1TCATTCTATCTATCGTGATG</span></p>
<p><span lang=EN-US>sequence_2TCATTCTATCTATCGTGATG</span></p>
<p><span lang=EN-US>sequence_3TCATTCTATCTATCGTGATG</span></p>
<p><span lang=EN-US>sequence_4TCATTCTATCTATCGTGATG</span></p>
<h2><a name="_Toc520201493"><span lang=EN-US>Does paml read nexus (PAUP) format?</span></a></h2>
<p>PAML reads nexus format if there are no comments in the sequences.  PAML reads 
  the sequence data only and ignores everything else.</p>
<h1><a name="_Toc520201494"><span lang=EN-US>Windows and UNIX basics</span></a></h1>
<p>Knowing how to drag a file icon might not be enough to use paml.  It would 
  be much easier if you know some basic MS Windows or UNIX commands or operations.  
  You should know how to create a folder (directory), change to a different folder, 
  copy files etc.  </p>
<p><span lang=EN-US>Here are a few basic commands for Windows: </span></p>
<table border=0 cellspacing=0 cellpadding=0>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>DOS/Windows</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>UNIX</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Function</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>cd</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>cd</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Change directory (folder)</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>md</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>mkdir</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Make a new directory</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>rd</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>rm -r </span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Remove a directory</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>copy </span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>cp</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Copy a file</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>ren </span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>mv</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Rename a file</span></p>
    </td>
  </tr>
  <tr> 
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>help</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>man</span></p>
    </td>
    <td valign=top class="Normal"> 
      <p><span lang=EN-US>Get help</span></p>
    </td>
  </tr>
</table>
<h1><a name="_Toc520201495"><span lang=EN-US>CODEML</span></a></h1>
<h2><a name="_Toc520201496"><span lang=EN-US>How can I get codeml to accept more 
  that one data set at a time?</span></a></h2>
<p>You use the control variable ndata.  Make sure that you remove the asterisk 
  *  at the beginning of the line, which means a comment line.<span lang=EN-US style='"Courier New"'></span></p>
<h2><a name="_Toc520201497"><span lang=EN-US>What are the limits in terms of number 
  of sequences in the alignment?</span></a></h2>
<p>The simple answer is a few hundred sequences for codon-based analysis (codonml), 
  one or two thousand sequences for nucleotide or amino acid based analysis (baseml 
  and aaml), if you have a good PC or workstation (as of June 2001).  The computation 
  increases more quickly with the number of sequences than with the number of 
  sites in the sequence.  How large data sets the programs can handle will depend 
  on many factors including the complexity of the model, how fast and how much 
  memory you have on your machine.  (Yang. 2000. J. Mol. Evol. 51:423-432)</p>
<h2><a name="_Toc520201498"><span lang=EN-US>Are my sequences too divergent for 
  a sensible ML analysis?</span></a></h2>
<p>In computer simulations, saturation of substitutions does not seem to be a 
  big problem for many real data sets, as the performance of the models does not 
  seem to deteriorate until the sequences become very divergent, say with 10 or 
  50 nucleotide substitutions per nucleotide site along the tree (Yang 1998. Syst. 
  Biol. 47:125-133; Anisimova et al. 2001. Mol. Biol. Evol. in press).  Also a 
  large tree with many branches will be able to tolerate more changes than a small 
  tree.  Similarly maximum likelihood joint analysis of all sequences is more 
  tolerant of multiple substitutions than pairwise distance methods.  However, 
  high sequence divergence is often associated with other problems, such as difficulty 
  in alignment, different codon usage biases or nucleotide compositions in different 
  sequences. </p>
<h2><a name="_Toc520201499"><span lang=EN-US>How do I test positive selection 
  along specific lineages?</span></a></h2>
<p>Codonml (codeml with seqtype = 1) has a few models that assign and estimate 
  different w (= dN/dS) ratios for different branches in the phylogeny.  The free-ratios 
  model (model = 1) fits one w ratio for each branch in the tree, and the two-ratio 
  or three-ratios models (model = 2) let you decide how many ratios you want and 
  which ratio each branch should have.  The program will then estimate those w 
  ratios by ML.  The reference is Yang (1998 Mol. Biol. Evol. 15:568-573).  The 
  example files are in the folder paml/examples/lysozyme/ in the package.  Also 
  look at the notes under “Branch labels” in the section “<a
name="_Toc508124877">Tree file and representations of tree topology</a>” in the 
  documentation.</p>
<h2><a name="_Toc520201500"><span lang=EN-US>Can I conclude positive selection 
  if I found significantly higher dN/dS ratios for my branches of interest than 
  for other branches?</span></a></h2>
<p style='text-autospace:none'>If your estimates of w for the branches of interest 
  are not &gt; 1, you might not be able to conclude positive selection based on 
  the likelihood ratio test alone.  Relaxed selective constraint along the lineages 
  of interest is an alternative compatible explanation.  Furthermore, under Ohta’s 
  hypothesis of slightly deleterious mutations, purifying selection is more effective 
  in large populations than in small populations, and so differences in population 
  sizes along lineages provide another compatible hypothesis.  If amino acid changes 
  are slightly deleterious, we expect them to be removed from the population at 
  a higher rate in a large population than in a small population.  As a result, 
  we expect to see a smaller dN/dS ratio in a large population than in a small 
  one, even if there is no difference between the two lineages in selective pressure 
  or gene function.  For example, the dN/dS ratios in many nuclear genes are lower 
  in rodents than in primates or artiodactyls (Ohta. 1995 J. Mol. Evol. 40:56-63; 
  Bielawski et al. 2000 Genetics 156:1299-1308; Yang &amp; Nielsen. 1998 J. Mol. 
  Evol. 46:409-418).</p>
<h2><a name="_Toc520201501"><span lang=EN-US>What is the effect of assuming the 
  same selective pressure among lineages?</span></a></h2>
<p>The lineage-based analysis assumes the same w (dN/dS) ratio among sites, and 
  detects selection only if the w averaged over all sites is greater than one.  
  This is expected to make the test conservative and lack power.  </p>
<h2><a name="_Toc520201502"><span lang=EN-US>Can I obtain a list of probabilities 
  for all sites and all categories under a NSsites model?</span></a></h2>
<p>Yes, the list is in the file rst.  I used such output to make the plot published 
  in Yang &amp; Bielawski (2000 TREE 15: 496-503).</p>
<h2><a name="_Toc520201503">The different NSsites models identified different 
  sites.  Which model should I believe?</a></h2>
<p>Different NSsites models such as M3 and M8 always produce different parameter 
  estimates and possibly different lists of sites under positive selection.  My 
  experience with real data indicates that the models are almost always consistent 
  about which sites might be under diversifying selection even if the lists are 
  different.  However, I have heard complaints about the fact that the lists are 
  not exactly the same.  There is no easy answer to this question.  The list in 
  the main output file is based on a cutoff of P = 50% and sites with P &gt; 95% 
  and P &gt; 99% are marked with * and ** respectively.  Typically if a site is 
  in the list under one model, it will have a substantial probability under another 
  model as well.  The results might not be so different if you view them in this 
  way.  The problem of identifying sites is difficult and is prone to errors, 
  partly because we are making so many inferences at the same time.  The situation 
  is similar to getting the top few students from a class.  The more you include 
  in your list, the poorer the quality. </p>
<h2><a name="_Toc520201504"><span lang=EN-US>Can I specify certain branches in 
  a tree file to have independent w parameters?</span></a></h2>
<p>Yes.  See the question “How do I test positive selection along specific lineages?”.</p>
<h2><a name="_Toc520201505"><span lang=EN-US>Can I specify certain sites in an 
  alignment to have independent w parameters?</span></a></h2>
<p>Yes, you can use prior information such as the protein 3-D structure to partition 
  sites (codons) in the gene into several different classes.  You can then specify 
  models that assign and estimate different w (dN/dS) parameters for the different 
  site partitions.  The models are implemented in the program in the framework 
  of analyzing multiple protein-coding genes for the same set of species, using 
  the G option of the sequence data file.  You then change the Mgene variable 
  in the control file codeml.ctl.  Look at “Option G” in the section “Sequence 
  data format” in the documentation.  The reference is Yang &amp; Swanson (2001 
  Mol. Biol. Evol. In press).</p>
<p>Note that these models are different from the NSsites models, which do not 
  require or make use of prior information to partition sites into classes.  The 
  NSsites models assume that different sites in the sequence have different w 
  ratios, but that we do not know which sites are conserved or which sites are 
  evolving fast.  The NSsites models add noise to the model by assuming a statistical 
  distribution (such as a general discrete distribution, M3, or beta, M7) to account 
  for the variability.</p>
<h2><a name="_Toc520201506"><span lang=EN-US>Can I estimate dN and dS between 
  all pairs of taxa in an alignment?</span></a></h2>
<p>Yes, you use runmode = -2.  The ML iteration is still slow, and each pairwise 
  comparison takes several seconds. </p>
<h2><a name="_Toc520201507"><span lang=EN-US>Why do the inferred site classes 
  not match the estimated proportions under the NSsites models?</span></a><span>  
  </span></h2>
<p>Thanks to Chris Woelk for asking the question originally.  Here are more details 
  about the question.  The parameter estimates under M2 (NSsites = 2) suggest 
  the following proportions and w (dN/ds) values:</p>
<p><span lang=EN-US>Frequencies for categories (K=3)</span></p>
<p><span>    </span>0.721330    0.000000    0.278670</p>
<p><span lang=EN-US>dN/dS rate ratios for categories (K=3)</span></p>
<p><span>    </span>0.000000    1.000000    0.156520</p>
<p>Yet when I counted up the number of sites assigned to Class 3, there were 86 
  of them.  86/517 = 0.166.  This is not equivalent to the frequency 0.278670 
  cited above.  </p>
<p>Answer: Those proportions are not expected to match, and your case is not extreme.  
  The statistical nature of inference of the site classes is the same as is the 
  following simple case.  Suppose I tell you that nine boys wear black hats and 
  one boy wears a white hat.  I then ask you to guess the colour of each of the 
  ten boys' hats.  Your best bet will be &quot;black&quot;, for each and every 
  boy.  The colour for each boy is black with 90% probability and white with 10% 
  probability.  Because we ignore the small probability of 10% in each case, the 
  combined result is an intrinsic &quot;bias&quot; in the exercise (I am not very 
  sure whether bias is the right word).  </p>
<p>There is nothing wrong with the exercise, as the probabilities are all correct 
  (given that your estimates of parameters are reliable).  And it is also the 
  best we can do, if people want to know the most likely site class for each site 
  (you can try to come out with a better solution to the hat colour problem I 
  posed.)  However, you should bear in mind that it is almost certain that the 
  inference at one or more sites is wrong, and also that you should not use the 
  inferred site classes to perform further statistical analysis (or otherwise 
  you need to bear in mind the bias).  For example if you use your &quot;inferrred&quot; 
  colours of hats to &quot;estimate&quot; the proportion of black hats, you get 
  10/10 rather than 9/10.</p>
<p>By the way, the calculation is done using the Bayes theorem, and the method 
  is known as the (empirical) Bayes method.  It is a proper and standard statistical 
  method.  Quite a few analyses in paml use it, and my comments above should apply 
  to all those cases.  Those include (1) the inference of the rate (or rate class) 
  for sites under models of variable substitution rates among sites (like the 
  gamma model), (2) inference of most likely ancestral nucleotides, amino acids, 
  or codons (that is, ancestral reconstruction).  Ancestral reconstruction has 
  been so popular and many people are using reconstructed characters as if they 
  were observed data, and so those methods are at least not proper.  Other uses 
  of the Bayes method include estimation of phylogenies, of alignment by Thorne 
  et al., of secondary structure categories for sites by Goldman et al.</p>
<p>Similarly the most likely rates for individual sites or to the most likely 
  ancestral character states predicted by the empirical Bayes method all involve 
  such systematic biases.  Ideally they should not be used as data to construct 
  statistical tests. </p>
<h1><a name="_Toc520201508"><span lang=EN-US>BASEML</span></a></h1>
<p><span lang=EN-US>To be written.</span></p>
<h1><a name="_Toc520201509"><span lang=EN-US>EVOLVER</span></a></h1>
<h2><a name="_Toc520201510"><span lang=EN-US>Where is the control file for evolver?</span></a></h2>
<p>Evolver does not have a control file as the other programs do.  You run the 
  program and it will display a naively simple menu.  If you choose to simulate 
  sequence data sets, the program will use data files MCbase.dat, MCcodon.dat, 
  and MCaa.dat to simulate nucleotide, codon, and amino acid sequences, respectively.  
  Print out a copy of the data file and change parameters as necessary.</p>
</body>
</html>
